PDF operators for each distribution

david-seiler opened the pull request:
## Description ##
This PR adds operators for computing the densities of samples drawn from any of the various distributions defined in operator/random, as well as their gradients.  There are lots of changes to test_random.py to test each PDF alongside its distribution; aside from that, the patch should be entirely stand-alone.  See pdf_op.cc for more-detailed description strings.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

asmushetzel commented on src/operator/random/pdf_op.h:
```diff
@@ -0,0 +1,594 @@
+  bool is_log;
+  DMLC_DECLARE_PARAMETER(PdfParam) {
+    DMLC_DECLARE_FIELD(is_log).set_default(false)
+    .describe("Whether to compute the log PDF or not.");
```
May be that can be phrased a bit better.

david-seiler replied:
It's now "If set, compute the density of the log-probability instead of the probability."

asmushetzel commented on src/operator/random/pdf_op.h:
```diff
@@ -0,0 +1,594 @@
+  template<typename DType, typename IType1, typename IType2>
+  MSHADOW_XINLINE static void Map(int start, int length, int sample_size, int index,
+                                  DType *out, IType1 *sample, IType2 *lower, IType2 *upper) {
+    const DType l(lower[index]), h(upper[index]);
```
Providing "index" to this and all similar functions below is redundant. According to the processing in LaunchExWrapper it always hold that index == start/sample_size. Still we may want to keep "index" as an explicit parameter for clarity. So o.k. with either choice.

david-seiler replied:
It's gone.

asmushetzel commented:
Should be noted that this PR allows to compute log-pdf and pdf. And that it also adds pdf of Dirichlet distribution which does not yet exist in MXNet. 
For consistency, we should then add also a sampler for Dirichlet-distributions after this PR gets merged.

asmushetzel commented:
This addresses several requests from issue #12932

anirudhacharya commented on src/operator/random/pdf_op.cc:
```diff
@@ -0,0 +1,292 @@
+)code");
+}
+
+inline std::string dirichlet_desc() {
```
I don't think we have a function to draw samples from a dirichlet distribution? this only computes the pdf of the distribution?

david-seiler replied:
Updated the commit message to clarify that, unlike the other distributions, we don't have a Dirichlet sampler yet.

marcoabreu commented:
Is this PR pointing against the 1.3 branch on purpose?

piyushghai commented:
@david-seiler Seems like you want to add a particular feature to a previous version of MXNet. 
The correct process to do this would be : 

- Raise a PR against the master branch. 
- After that PR is merged into master, cherry-pick the same commit in another PR raised against v1.3.x (Or a different version) branch. 

@mxnet-label-bot Add [pr-awaiting-review, operator]

david-seiler commented:
> @david-seiler Seems like you want to add a particular feature to a previous version of MXNet.

I started with 1.3.x because it's what we're using internally (for now); I'll close this in favor of a new PR against master once I've updated to build against master and addressed the rest of the review comments.

david-seiler commented:
This PR now correctly targets master -- still from a branch named v1.3.x on my side, since I couldn't see how to reset that, but that branch now follows apache:master with this one commit on top.  Sorry for the confusion, but it seems better on net than throwing away the existing conversation.

david-seiler commented on src/operator/random/pdf_op.cc:
```diff
@@ -0,0 +1,292 @@
+
+  random_pdf_exponential(sample=[[1, 2, 3]], lam=[1]) =
+      [[0.36787945 0.13533528 0.04978707]]
+
```
Added second examples for the exponential and Dirichlet.

david-seiler commented on src/operator/random/pdf_op.cc:
```diff
@@ -0,0 +1,292 @@
+produces output of dimension *n+1* such that each *n*-dimensional index *i*
+in the output array holds the PDFs of the samples at index *i* in *sample*,
+parameterized by the values of *low* and *high* at index *i*.
+
```
Called that out explicitly.

david-seiler commented:
The tests in Jenkins have been stuck ever single I switched this PR to master, and I haven't found a way to unstick them, so please join me in a new PR of the same code: https://github.com/apache/incubator-mxnet/pull/14617

All the comments from this PR should be addressed there.
